### PR TITLE
Allow threads

### PR DESCRIPTION
**Pull request by adriangb**
This should release the GIL and allow use in multiple threads.

I tested with this script:

```python
from contextlib import contextmanager
import math
from time import time
from typing import Iterator

import anyio
import anyio.to_process
import anyio.to_thread
import object_store


@contextmanager
def timeit(name: str) -> Iterator[None]:
    start = time()
    yield
    print(f'{name} took {time() - start:.2f} seconds', flush=True)


def work() -> None:
    object_store.ObjectStore('gs://adriangb-public-bucket').get('yellow_tripdata_2024-01 (1).parquet')


async def awork(limiter: anyio.CapacityLimiter) -> None:
    await anyio.to_thread.run_sync(work, limiter=limiter)


async def main() -> None:
    limiter = anyio.CapacityLimiter(math.inf)

    with timeit('main'):
        async with anyio.create_task_group() as tg:
            for _ in range(32):
                tg.start_soon(awork, limiter)


if __name__ == '__main__':
    anyio.run(main)
```

Locally there isn't much difference, I'm IO bound. But on GCP compute that goes from ~15s to ~3s for me.

### CODE REVIEW

File: object-store-internal/src/lib.rs [hunk 1]
```diff
@@ -507,73 +508,91 @@ impl PyObjectStore {
...
-            .block_on(self.inner.put(&location.into(), bytes.into()))
-            .map_err(ObjectStoreError::from)?;
-        Ok(())
+    fn put(&self, py: Python, location: PyPath, bytes: Vec<u8>) -> PyResult<()> {
+        py.allow_threads(|| {
+            self.rt
+                .block_on(self.inner.put(&location.into(), bytes.into()))
+                .map_err(ObjectStoreError::from)?;
+            Ok(())
+        })
```
Comment (kylebarron) [inline object-store-internal/src/lib.rs:517]: I always thought that if you didn't include the `py: Python` argument, that the GIL would be released automatically. Is this not true? To release the GIL you have to include the `py` argument **and** then also call `allow_threads`?

Comment (adriangb) [reply object-store-internal/src/lib.rs:517]: I thought the same thing but that's not what my test showed. If you look at the docs: https://pyo3.rs/v0.21.1/parallelism#parallelism

It goes from

```rust
fn search_sequential(contents: &str, needle: &str) -> usize {
    contents.lines().map(|line| count_line(line, needle)).sum()
}
```

To

```rust
fn search_sequential_allow_threads(py: Python<'_>, contents: &str, needle: &str) -> usize {
    py.allow_threads(|| search_sequential(contents, needle))
}
```

So I think we are wrong.

I can confirm with @davidhewitt tomorrow.

Comment (davidhewitt) [reply object-store-internal/src/lib.rs:517]: Indeed the GIL is not released automatically, as many arguments like `Bound<'py, PyAny>` can also implicitly carry the GIL even if `py: Python<'py>` is not included.

**Comment by adriangb** [comment]
@roeap quick ping on this

**Comment by roeap** [comment]
@adriangb - sorry for being MIA for so long. Do you mind rebasing, and I am happy to review / merge then.

**Comment by ion-elgreco** [comment]
@roeap will you push a 0.2.0 release after this one?

**Pull request merged**